src/rgw/driver/d4n/d4n_policy.cc

```diff
@@ -473,175 +575,216 @@ void LFUDAPolicy::cleaning(const DoutPrefixProvider* dpp)
+	  ldpp_dout(dpp, 20) << __func__ << "(): oid_in_cache =" << oid_in_cache << dendl;
+	  std::string new_oid_in_cache = prefix + CACHE_DELIM + std::to_string(fst) + CACHE_DELIM + std::to_string(cur_len);
+	  //Rename block to remove "D" prefix
+	  cacheDriver->rename(dpp, oid_in_cache, new_oid_in_cache, null_yield);
+	  //Update in-memory data structure for each block
+	  this->update(dpp, new_oid_in_cache, 0, 0, e->version, false, y);
+
+	  rgw::d4n::CacheBlock block;
+	  block.cacheObj.bucketName = c_obj->get_bucket()->get_bucket_id();
+	  block.cacheObj.objName = c_obj->get_key().get_oid();
+	  block.size = cur_len;
+	  block.blockID = fst;
+	  op_ret = blockDir->update_field(dpp, &block, "dirty", "false", null_yield);
+	  if (op_ret < 0) {
+	    ldpp_dout(dpp, 0) << __func__ << "updating dirty flag in block directory failed, ret=" << op_ret << dendl;
+	  }
+	  fst += cur_len;
+	} while(fst < lst);
+
+	cacheDriver->rename(dpp, head_oid_in_cache, new_head_oid_in_cache, null_yield);
+	//data is clean now, updating in-memory metadata for an object
+	e->dirty = false;
+	//invoke update() with dirty flag set to false, to update in-memory metadata for head
+	this->update(dpp, new_head_oid_in_cache, 0, 0, e->version, false, y);
+
+	rgw::d4n::CacheBlock block;
+	block.cacheObj.bucketName = c_obj->get_bucket()->get_bucket_id();
+	block.cacheObj.objName = c_obj->get_name();
+	block.size = 0;
+	block.blockID = 0;
+	if (c_obj->have_instance()) {
+	  blockDir->get(dpp, &block, null_yield);
+	  if (block.version == c_obj->get_instance()) { //versioned case - update head block entry that has latest version
+	    op_ret = blockDir->update_field(dpp, &block, "dirty", "false", null_yield);
+	    if (op_ret < 0) {
+		ldpp_dout(dpp, 20) << __func__ << "updating dirty flag in block directory for head failed!" << dendl;
+	    }
+	  }
+	} else { //non-versioned case
+	  op_ret = blockDir->update_field(dpp, &block, "dirty", "false", null_yield);
+	  if (op_ret < 0) {
+	      ldpp_dout(dpp, 20) << __func__ << "updating dirty flag in block directory for head failed!" << dendl;
+	  }
+	}
+	if (c_obj->have_instance()) {
+	  rgw::d4n::CacheBlock instance_block;
+	  instance_block.cacheObj.bucketName = c_obj->get_bucket()->get_bucket_id();
+	  instance_block.cacheObj.objName = c_obj->get_oid();
+	  instance_block.size = 0;
+	  instance_block.blockID = 0;
+	  op_ret = blockDir->update_field(dpp, &instance_block, "dirty", "false", null_yield);
+	  if (op_ret < 0) {
+	      ldpp_dout(dpp, 20) << __func__ << "updating dirty flag in block directory for instance block failed!" << dendl;
+	  }
+	}
+	//remove entry from map and queue, erase_dirty_object locks correctly
+	erase_dirty_object(dpp, e->key, null_yield);
+      } else { //end-if std::difftime(time(NULL), e->creationTime) > interval
+	std::this_thread::sleep_for(std::chrono::milliseconds(interval)); //TODO:: should this time be optimised?
```

Comment:
why is the thread put to sleep  here and on line 787

Reply:
On line 783, if interval seconds hasn't passed yet, then the thread must be put to sleep. On line 786, I also added the `sleep_for` call if the key is empty but the entry is in the heap for some reason. You previously stated that the loop should not continue in this case, so the other option is to do nothing. Is doing nothing better here than putting the thread to sleep again?

Reply:
If the key is empty and you put the thread to sleep, even when it wakes up, it will still see the entry with the empty key correct? and then the thread would be put to sleep again, and this might continue forever.

Reply:
A better option is to just check if invalid is true and then do the steps for invalid blocks. Then club the key empty check with time expiration check and put the thread to sleep when the time has not expired. Ideally the key will never be empty.